### PR TITLE
Add an experimental plugin to update some new Wacom tablets

### PR DESCRIPTION
**Pull request by hughsie**
This is still WIP. The touch module firmware update hasn't been tested yet, but the other firmware types work fine. I figured I'd get this in for some early review as it's quite a chunk of code.

### CODE REVIEW

File: plugins/wacomhid/fu-self-test.c [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+#include "fu-common.h"
+#include "fu-test.h"
+#include "fu-wac-common.h"
+#include "fu-wac-firmware.h"
+
+#include "fwupd-error.h"
+
+static void
+fu_firmware_wac_func (void)
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-self-test.c]: Fixed.

File: plugins/wacomhid/fu-self-test.c [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	g_test_init (&argc, &argv, NULL);
+
+	/* only critical and error are fatal */
+	g_log_set_fatal_mask (NULL, G_LOG_LEVEL_ERROR | G_LOG_LEVEL_CRITICAL);
+
+	/* log everything */
+	g_setenv ("G_MESSAGES_DEBUG", "all", FALSE);
+
+	/* tests go here */
+	g_test_add_func ("/wac/firmware", fu_firmware_wac_func);
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-self-test.c]: Fixed.

File: plugins/wacomhid/fu-wac-device.c [hunk 3]
```diff
@@ -0,0 +1,925 @@
...
+	/* reboot, which switches the boot index of the firmware */
+	return fu_wac_device_update_reset (self, error);
+}
+
+static gboolean
+fu_wac_device_probe (FuUsbDevice *device, GError **error)
+{
+	const gchar *plugin_hints;
+
+	/* devices have to be whitelisted */
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-wac-device.c]: I was slightly anxious that probing the device might actually do bad things. I'll ask Wacom for the best plan here.

File: plugins/wacomhid/fu-wac-common.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+G_BEGIN_DECLS
+
+#define HID_REPORT_GET					0x01
+#define HID_REPORT_SET					0x09
+
+#define HID_REPORT_TYPE_INPUT				0x01
+#define HID_REPORT_TYPE_OUTPUT				0x02
+#define HID_REPORT_TYPE_FEATURE				0x03
+
+#define HID_FEATURE					0x0300
```
[the remark this answers is not in the document]
Comment (superm1) [reply plugins/wacomhid/fu-wac-common.h]: Seems like a good enough place, yep.

File: plugins/wacomhid/fu-wac-device.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-wac-device.c]: Wacom are going to get back to me about this.

File: plugins/wacomhid/fu-wac-device.c [hunk 5]
```diff
@@ -0,0 +1,925 @@
...
+	/* hit hardware */
+	return fu_wac_device_set_feature_report (self, buf, bufsz,
+						 FU_WAC_DEVICE_FEATURE_FLAG_NONE,
+						 error);
+}
+
+static gboolean
+fu_wac_device_erase_block (FuWacDevice *self, guint32 addr, GError **error)
+{
+	guint8 buf[5];
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-wac-device.c]: Fixed.

File: plugins/wacomhid/fu-wac-device.c [hunk 6]
```diff
@@ -0,0 +1,925 @@
...
+	fu_device_add_flag (FU_DEVICE (device), FWUPD_DEVICE_FLAG_UPDATABLE);
+	return TRUE;
+}
+
+static gboolean
+fu_wac_device_add_modules_bluetooth (FuWacDevice *self, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	const gsize sz = 15;
+	guint8 buf[sz];
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/wacomhid/fu-wac-device.c]: Fixed.

**Comment by hughsie** [comment]
Lots of fixes later, I'd like to get this on master ASAP (even if there's no LVFS firmware yet) if nothing else to be able to use Coverity and friends. I've been testing this on my tablet updating and downgrading both the main and bluetooth firmware and it all seems to work. Any reservations?

**Pull request merged**